ostelco-ios-client/ViewControllers/Home/ApplePayViewController.swift

```diff
@@ -0,0 +1,80 @@
+//
+//  ApplePayViewController.swift
+//  ostelco-ios-client
+//
+//  Created by Prasanth Ullattil on 26/04/2019.
+//  Copyright © 2019 mac. All rights reserved.
+//
+
+import UIKit
+import Stripe
+import Siesta
+
+// Default implementaion for the PKPaymentAuthorizationViewControllerDelegate methods.
+class ApplePayViewController: UIViewController, ApplePayDelegate {
```

Comment:
Is it possible to make this a protocol restricted to `UIViewController`? You can see an example of this with [`LocationCheckingViewController`](https://github.com/ostelco/ostelco-ios-client/blob/develop/ostelco-ios-client/Protocols/LocationCheckingViewController.swift), but it essentially allows us to have opt-in functionality without having to subclass.

Reply:
I tried, but was hitting the problem listed here. https://stackoverflow.com/questions/39487168/non-objc-method-does-not-satisfy-optional-requirement-of-objc-protocol

Reply:
Ah yeah, so you can't just conform to a protocol with a protocol extension.  

There's a couple different workarounds to this but what I normally do is still implement the UIKit protocol on the view controller but just have it call through to a protocol extension I've defined. For example: 

```swift
/// MyProtocol.swift

// Protocol declaration
protocol MyProtocol {
   func handleAppleDidSomething() 
} 

// Default implementation of protocol
extension MyProtocol {
   func handleAppleDidSomething() {
      // Common handling
   }
}

/// MyViewController.swift

// Declaration of class
class MyViewController: UIViewController {
   // View controllery things
}

// Mix-in extension since we just want the behavior of the default implementation
extension MyViewController: MyProtocol { }

// Extension conforming to apple protocol that can't deal with @nonobjc things
extension MyViewController: AppleProtocol {
     func appleDidSomething() {
       self.handleAppleDidSomething()
    }
}
```

You also could constrain your default extension to only apply if the conforming object also implements `AppleProtocol`: 

```swift
extension MyProtocol where Self: AppleProtocol {
```

That way if the thing conforming to `MyProtocol` view controller doesn't also implement `AppleProtocol`, it doesn't get the default implementation for free.

Reply:
Thanks for the help. I have still retained the `ApplePayViewController` base class. That implements the `ApplePayDelegate` properties and `PKPaymentAuthorizationViewControllerDelegate` methods